### DISCUSSION TITLE
Adapters to other popular frameworks' "Connection" equivilents

### DISCUSSION THREAD

**Discussion by cylewitruk** [diesel-rs/diesel#3938] [Ideas]
As I've mentioned in other posts/Q's I'm trying to retrofit Diesel into a larger app which makes heavy use of 'rusqlite' kinda all over the place. One thing that would make this a lot easier is a way to use rusqlite's `Connection` as a backing connection in diesel, so that an open connection could be attached to and refactored, to avoid having to change everything at once.

I've looked a little at how new connection's are written, but I'm curious if there's an easier way than basically copying the entire sqlite connection and replacing all low-level calls with calls via rusqlite?

**Comment by weiznich** [comment]
Generally speaking it's possible to implement `Connection` outside of diesel for an existing implementation of `Backend`. That's something that's explicitly supported. For postgres and mysql there is a clear distinction between what's backend related and what's related to the `Connection` impl in diesel. There you can implement `Connection` outside of diesel to use different underlying connection types (as long as they allow to access various low level functionality).

Now for sqlite the situation is more complicated. Due to how sqlite work and how diesel evolved over the years we sticked with access by libsqlite3-sys directly instead of using rusqlite. That means we access quite a lot of the really low level details to perform various optimizations and to get things right. That makes it really hard to replace the existing implementation with something else. That written I'm not generally opposed in accepting PR's that slowly change it to a rusqlite based approach as long as that does not regress performance and as long as that does not couple our public API to their public API.
